### ISSUE TITLE
Error "ValueError: No arrays to thread error on to" when vectorizing over diffrax eqsolve

### ISSUE THREAD

**Issue by dpanici** [patrick-kidger/diffrax#292]
I don't yet have a MWE as this only seems to occur when using a script I have in my own package, but I am wondering if some light can be shed on what exactly this error means? Sorry as well for the large amount of code dumped here, I don't yet have a MWE but maybe if this error has been seen before I can then figure out why it is occurring

I see this originates in Equinox but am unsure of to put the issue here or there. For context when I run pdb here and check the values of the arguments to the errorif function I find they are:

```
(Pdb) down
> /home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/diffrax/integrate.py(542)diffeqsolve()
-> dt0 = eqxi.error_if(dt0, pred, msg)
(Pdb) print(dt0)
3.141592653589793
(Pdb) print(pred)
Traced<ShapedArray(bool[])>with<DynamicJaxprTrace(level=1/0)>
(Pdb) print(msg)
Must have (t1 - t0) * dt0 >= 0, we instead got t1 with value Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=1/0)> and type <class 'jax._src.interpreters.partial_eval.DynamicJaxprTracer'>, t0 with value Traced<ShapedArray(float64[])>with<DynamicJaxprTrace(level=1/0)> and type <class 'jax._src.interpreters.partial_eval.DynamicJaxprTracer'>, dt0 with value 3.141592653589793 and type <class 'float'>
```

The full traceback is:

```python
Traceback (most recent call last):
  File "/home/dpanici/DESC/work/random/diffrax/script_for_slow_diffrax_comopile.py", line 32, in <module>
    field_line_integrate(r0, z0, phis, field,min_step_size=np.pi)
  File "/home/dpanici/DESC/desc/magnetic_fields.py", line 911, in field_line_integrate
    x = jnp.vectorize(intfun, signature="(k)->(n,k)")(x0)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/jax/_src/numpy/vectorize.py", line 309, in wrapped
    result = vectorized_func(*squeezed_args)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/jax/_src/numpy/vectorize.py", line 136, in wrapped
    out = func(*args)
          ^^^^^^^^^^^
  File "/home/dpanici/DESC/desc/magnetic_fields.py", line 897, in <lambda>
    intfun = lambda x: diffeqsolve(
                       ^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/jax/_src/traceback_util.py", line 166, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/jax/_src/pjit.py", line 249, in cache_miss
    outs, out_flat, out_tree, args_flat = _python_pjit_helper(
                                          ^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/jax/_src/pjit.py", line 155, in _python_pjit_helper
    args_flat, _, params, in_tree, out_tree, _ = infer_params_fn(
                                                 ^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/jax/_src/api.py", line 301, in infer_params
    return pjit.common_infer_params(pjit_info_args, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/jax/_src/pjit.py", line 498, in common_infer_params
    jaxpr, consts, canonicalized_out_shardings_flat = _pjit_jaxpr(
                                                      ^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/jax/_src/pjit.py", line 960, in _pjit_jaxpr
    jaxpr, final_consts, out_type = _create_pjit_jaxpr(
                                    ^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/jax/_src/linear_util.py", line 345, in memoized_fun
    ans = call(fun, *args)
          ^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/jax/_src/pjit.py", line 913, in _create_pjit_jaxpr
    jaxpr, global_out_avals, consts = pe.trace_to_jaxpr_dynamic(
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/jax/_src/profiler.py", line 314, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/jax/_src/interpreters/partial_eval.py", line 2150, in trace_to_jaxpr_dynamic
    jaxpr, out_avals, consts = trace_to_subjaxpr_dynamic(
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/jax/_src/interpreters/partial_eval.py", line 2172, in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers_)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/jax/_src/linear_util.py", line 188, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/equinox/_jit.py", line 41, in fun_wrapped
    out = fun(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/diffrax/integrate.py", line 542, in diffeqsolve
    dt0 = eqxi.error_if(dt0, pred, msg)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/equinox/_errors.py", line 204, in error_if
    return branched_error_if(x, pred, 0, [msg], on_error=on_error)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/equinox/_errors.py", line 264, in branched_error_if
    raise ValueError("No arrays to thread error on to.")
jax._src.traceback_util.UnfilteredStackTrace: ValueError: No arrays to thread error on to.

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/dpanici/DESC/work/random/diffrax/script_for_slow_diffrax_comopile.py", line 32, in <module>
    field_line_integrate(r0, z0, phis, field,min_step_size=np.pi)
  File "/home/dpanici/DESC/desc/magnetic_fields.py", line 911, in field_line_integrate
    x = jnp.vectorize(intfun, signature="(k)->(n,k)")(x0)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/jax/_src/numpy/vectorize.py", line 309, in wrapped
    result = vectorized_func(*squeezed_args)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/jax/_src/numpy/vectorize.py", line 136, in wrapped
    out = func(*args)
          ^^^^^^^^^^^
  File "/home/dpanici/DESC/desc/magnetic_fields.py", line 897, in <lambda>
    intfun = lambda x: diffeqsolve(
                       ^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/equinox/_jit.py", line 107, in __call__
    return self._call(False, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/equinox/_jit.py", line 103, in _call
    out = self._cached(dynamic, static)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dpanici/miniconda3/envs/desc-env/lib/python3.11/site-packages/diffrax/integrate.py", line 542, in diffeqsolve
    dt0 = eqxi.error_if(dt0, pred, msg)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: No arrays to thread error on to
```

And the function block in my script is

```python
def field_line_integrate(
    r0,
    z0,
    phis,
    field,
    params={},
    rtol=1e-8,
    atol=1e-8,
    maxstep=1000,
    min_step_size=1e-8,
    solver=Tsit5(),
    terminating_event=None,
    bounds_R=(-np.inf, np.inf),
    bounds_Z=(-np.inf, np.inf),
    bounds_phi=(-np.inf, np.inf),
    kwargs={},
):
    """Trace field lines by integration, using diffrax package.

    Parameters
    ----------
    r0, z0 : array-like
        initial starting coordinates for r,z on phi=phis[0] plane
    phis : array-like
        strictly increasing array of toroidal angles to output r,z at
        Note that phis is the geometric toroidal angle for positive Bphi,
        and the negative toroidal angle for negative Bphi
    field : MagneticField
        source of magnetic field to integrate
    params: dict
        parameters passed to field
    rtol, atol : float
        relative and absolute tolerances for ode integration
    maxstep : int
        maximum number of steps between different phis
    min_step_size: float
        minimum step size (in phi) that the integration can take. default is 1e-8
    solver: diffrax.Solver
        diffrax Solver object to use in integration,
        defaults to Tsit5(), a RK45 explicit solver
    terminating_event_fxn: fxn
        Function which takes as input the state of the ODE solution at each timestep
        and **kwargs, and outputs a bool which, if True, terminates the solve at that
        timestep.
        fxn must have signature of (state, **kwargs) -> Bool
        NOTE: If the solve is terminated early, the output returned is still
        length(phis), however all values from the step point the fxn evaluated
        to True and on will be inf
        state has attributes such as state.y (array of length 3 with current (R,phi,Z))
        see diffrax documentation for more in-depth information
    bounds_R: tuple
        tuple of (R_min,R_max) of the R bounds for the domain of interest.
        Integration will terminate when the field line exits this domain
        (when using the default terminating event)
    bounds_Z: tuple
        tuple of (Z_min,Z_max) of the Z bounds for the domain of interest.
        Integration will terminate when the field line exits this domain
        (when using the default terminating event)
    bounds_phi: tuple
        tuple of (phi_min,phi_max) of the phi bounds for the domain of interest.
        Integration will terminate when the field line exits this domain
        (when using the default terminating event)

    kwargs: dict
        keyword arguments to be passed into the diffrax diffeqsolve function call



    Returns
    -------
    r, z : ndarray
        arrays of r, z coordinates at specified phi angles

    """
    if not isinstance(solver, AbstractSolver):
        try:
            # maybe they passed in the correct object but did not
            # instantiate it
            solver = solver()
        except TypeError:
            raise TypeError(
                "Expected a diffrax Solver object as solver,"
                + f"instead got type {type(solver)}!"
            )
    r0, z0, phis = map(jnp.asarray, (r0, z0, phis))
    assert r0.shape == z0.shape, "r0 and z0 must have the same shape"
    rshape = r0.shape
    r0 = r0.flatten()
    z0 = z0.flatten()
    x0 = jnp.array([r0, phis[0] * jnp.ones_like(r0), z0]).T

    @jit
    def odefun(s, rpz, args):
        rpz = rpz.reshape((3, -1)).T
        r = rpz[:, 0]
        br, bp, bz = field.compute_magnetic_field(rpz, params, basis="rpz").T
        return jnp.array(
            [r * br / bp * jnp.sign(bp), jnp.sign(bp), r * bz / bp * jnp.sign(bp)]
        ).squeeze()

    # diffrax parameters
    stepsize_controller = PIDController(rtol=rtol, atol=atol, dtmin=min_step_size)

    def default_terminating_event_fxn(state, **kwargs):
        R_out_of_bounds = jnp.any(
            jnp.array(
                [
                    jnp.less(state.y[0], bounds_R[0]),
                    jnp.greater(state.y[0], bounds_R[1]),
                ]
            )
        )
        Z_out_of_bounds = jnp.any(
            jnp.array(
                [
                    jnp.less(state.y[2], bounds_Z[0]),
                    jnp.greater(state.y[2], bounds_Z[1]),
                ]
            )
        )
        phi_out_of_bounds = jnp.any(
            jnp.array(
                [
                    jnp.less(state.y[1], bounds_phi[0]),
                    jnp.greater(state.y[1], bounds_phi[1]),
                ]
            )
        )

        return jnp.any(jnp.array([R_out_of_bounds, Z_out_of_bounds, phi_out_of_bounds]))

    terminating_event = (
        DiscreteTerminatingEvent(terminating_event)
        if terminating_event
        else DiscreteTerminatingEvent(default_terminating_event_fxn)
    )
    term = ODETerm(odefun)
    saveat = kwargs.get("saveat", SaveAt(ts=phis))

    intfun = lambda x: diffeqsolve(
        term,
        solver,
        y0=x,
        t0=phis[0],
        t1=phis[-1],
        saveat=saveat,
        max_steps=maxstep,
        dt0=min_step_size,
        stepsize_controller=stepsize_controller,
        discrete_terminating_event=terminating_event,
        **kwargs,
    ).ys

    x = jnp.vectorize(intfun, signature="(k)->(n,k)")(x0)
    r = x[:, :, 0].squeeze().T.reshape((len(phis), *rshape))
    z = x[:, :, 2].squeeze().T.reshape((len(phis), *rshape))

    return r, z
```

**Comment by patrick-kidger** [comment]
What version of Diffrax are you using? This looks like a bug that was fixed in v0.4.1. (The latest release.)

**Comment by dpanici** [comment]
Ah I had 0.3.1, with 0.4.1 this issue is solved, sorry about that, thanks for the help!
